demo-orgs: Create demo organization without setting an email.

laurynmm opened the pull request:
Revision of the 1st commit in #19741, based on [Tim's review comment here](https://github.com/zulip/zulip/pull/19741/commits/f662b0834cb157e4ba161534ee88290a28fa55c3#r932756049).

Working specifically towards this part of #19523:
- We intend to allow you to initially create a demo organization without setting an email address, to make it feel super lightweight.

---

**Notes**:
- **1st commit**: Implements Tim's plan for not queuing welcome emails when the realm for the demo organization is being created since the owner's email is an empty string.

- **2nd commit**: I noted that without an email, key parts of the web-app in the new demo organization did not work, such as sending private messages to the welcome bot or to the new owner/user themself. I thought one option would be creating the realm with the email visibility already restricted, which is what this commit implements. But this may not be the route we want to take, which is why I kept it in a separate commit for now.

- **3rd commit**: I noted when doing some light manual testing in the development environment that the server was trying to send new login emails if I logged out/in of the new demo organization. I went ahead and added a similar restriction to 
`email_on_new_login` as in the first commit for `enqueue_welcome_emails`. I kept this as a separate commit in case there was some other implications for this change that I haven't yet considered.


**Questions**:

- In `accounts_register`, we validate email before doing much else, which we need to skip for demo organizations. I put in a check / assertion for the email being an empty string and `realm_creation` being true to skip the validation since that's how (at this point) we would ascertain that this is a demo organization. Is that enough to skip the validation? Or do we want a clearer marker that a demo organization is being created?

- The log-in page for the development environment uses the delivery email for input element value, so the button for the new owner of the demo realm is blank (see screenshot below) as it's an empty string. Maybe we could add labels with the user's full name for the owners? Or not worry about it?

--- 

**Screenshots and screen captures:**

<details>
<summary>Dev log-in screen for demo organization</summary>

### Note that because the delivery email is an empty string, there is no name/info in the button
![Screenshot from 2022-08-05 20-09-40](https://user-images.githubusercontent.com/63245456/183136755-f090cbb1-432d-4404-a5dd-cb148c48a307.png)
</details>

---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

laurynmm commented on zerver/views/development/registration.py:
```diff
@@ -87,8 +87,9 @@ def register_development_realm(request: HttpRequest) -> HttpResponse:
     count = UserProfile.objects.count()
-    name = f"user-{count}"
-    email = f"{name}@zulip.com"
+    name = f"demo-admin-{count}"
```
I felt like it would help to differentiate these users from ones created in the normal new realm creation flow.

timabbott replied:
Sure, that seems very reasonable.

laurynmm commented on zerver/views/registration.py:
```diff
@@ -170,10 +170,13 @@ def accounts_register(
-    except ValidationError:
-        return render(request, "zerver/invalid_email.html", context={"invalid_email": True})
+    if len(email) == 0:
+        assert realm_creation
```
Because the email is an empty string we need to skip the email validation. I put in an assertion for realm creation being true, but I'm not sure if it would make sense to try and get the `is_demo_organization` data from the request here as well.

timabbott replied:
We should definitely also assert `is_demo_organization` here as well.

laurynmm replied:
Continued this discussion [here](https://github.com/zulip/zulip/pull/22666/files#r949544828) with updated code.

timabbott commented:
> The log-in page for the development environment uses the delivery email for input element value, so the button for the new owner of the demo realm is blank (see screenshot below) as it's an empty string. Maybe we could add labels with the user's full name for the owners?

Yeah, we should put something in there; probably name is most reasonable, or user ID if that were to not be available.

laurynmm commented on zerver/views/registration.py:
```diff
@@ -171,10 +172,26 @@ def accounts_register(
+        #
+        # TODO: Remove settings.DEVELOPMENT when demo organizations feature ready
+        # to be fully implemented.
+        assert realm_creation and demo_organization_creation and settings.DEVELOPMENT
```
Revised version building on [this comment](https://github.com/zulip/zulip/pull/22666#discussion_r939130603). This is where Eeshan got held up in the initial PR with these changes (see discussion [here](https://github.com/zulip/zulip/pull/19741#discussion_r788252405) and [here](https://github.com/zulip/zulip/pull/19741/commits/f662b0834cb157e4ba161534ee88290a28fa55c3#r932756780).

This doesn't feel like a good solution, but I thought it worth getting some feedback.

Asserting that `settings.DEVELOPMENT` is true:
- Is this the best way to do that?
- Do we feel like that limitation combined with `realm_creation` is enough for now, and that we can revisit the `req_is_demo_organization` string or boolean parameter later in the feature development process?
- Or do we feel like that limitation is not needed at all?

`demo_organization_creation`:
- I decided to process through `REQ` and not do a `request.POST.get("is_demo_organization").
- The flow of this view does seem to require the email validation happening before the form creation, so again we need to figure out the converting / validating the string parameter to a boolean.

timabbott replied:
I might do `assert settings.DEVELOPMENT` on its own line so it's easier to give it a separate comment from the other assertions.

I think it's very helpful to protect an unfinished overall feature by limiting it to the development environment; it means we don't need to think about things like "Is it a really bad experience if someone runs one of these before X is done?".

timabbott replied:
`req_is_demo_organization` is a fine name; I think our usual convention is to use something like `_raw` if it's the right type but needs some processing; perhaps `is_demo_organization_string` would be the best name I can think of?

laurynmm replied:
@timabbott - Thanks for this feedback! I've moved `assert settings.DEVELOPMENT` to its own line. And I updated to use the suggested `is_demo_organization_string` for the parameter name as well.

laurynmm commented:
**Updates**:
- Revised the logic and assertion for not validating an email that's an empty string. See [this comment](https://github.com/zulip/zulip/pull/22666/files#r949544828).
- Added a 4th commit that addresses the empty button text in the dev login page when the email address for the owner is an empty string. See updated screenshot below. I only updated the demo organization owner buttons to use the user name; all other buttons are still user emails. But we could also just update all of the dev log-in buttons to show the user name vs email.

Including a GIF of basic navigation with all commits running in the dev environment.

**Note**: The first commit passes CI, but due to the email not being set and the email visibility being for all users, the app won't load. The second commit addresses that issue, but again that may not be the route we want to take as business users/organizations will likely want that setting to be the normal default. Though to invite other users to the demo organization, the owner will have to provide an email, so maybe when we do that flow, we could ask about whether they want to update that setting? Or update it to the default without them knowing?

<details>
<summary>Updated screenshots of dev login page with these updates</summary>

![Screenshot from 2022-08-18 16-58-34](https://user-images.githubusercontent.com/63245456/185485522-1feea852-95fb-4ac3-9727-07f0a0e01be6.png)
![Screenshot from 2022-08-18 16-59-06](https://user-images.githubusercontent.com/63245456/185485523-8a03d484-fde0-4174-9b59-0ad4df3535dd.png)
</details>


<details>
<summary>GIF of creation, sending messages and logging in/out of the dev environment</summary>

![Peek 2022-08-18 21-33](https://user-images.githubusercontent.com/63245456/185485556-84e76254-110a-49a3-9488-8ddd3c521a58.gif)

</details>

alya commented:
Nice! I have a couple of comments after playing around with this a bit.

alya commented:
First, let's pick a friendlier default name. How about just "Your name"?

alya commented:
The other comment is that the email and password editing fields look a bit confusing:

<img width="302" alt="Screen Shot 2022-08-22 at 2 50 23 PM" src="https://user-images.githubusercontent.com/2090066/186025508-dd3b30a9-bb46-4d8b-ad7b-5b1bd5dee30e.png">

1. Instead of showing an empty editable email field, maybe we should show a "Set email" link?
2. It looks like there's a password set, even though the user hasn't actually set one... Should we just not show the password field until an email has been configured? Similarly to how it's hidden in orgs that don't allow password login:

<img width="349" alt="Screen Shot 2022-08-22 at 2 57 10 PM" src="https://user-images.githubusercontent.com/2090066/186026037-7a7951ce-d4b5-4562-8991-9a0eea64ab9b.png">

3. I don't actually understand how the user is supposed to set their password, as I'm asked for my old password when I try to change it... How does this flow work?

laurynmm commented:
**Updates**:
- The default name for the demo organization owner is "Your name", see 1st screenshot below.
- In the personal account settings tab:
  - The password field is hidden for demo organization owners until they set their email address to not be blank.
  - Instead of the normal email field, there is a button for the owner until the email address is not blank, see 2nd screenshot below.
  - Copy of change email form as a template for the set email form for demo organizations. Note that form does not currently do anything, see to-do notes.

---

**To-do next / pending**:
- Clarifying the form and flow for adding an email (and password?).
- Deciding if any settings / actions other than inviting other users needs to be restricted until the demo organization owner updates for their email and password to log in.

---

**Screenshots**:

<details>
<summary>Screen after creating a demo organization in dev environment</summary>

![Screenshot from 2022-11-22 20-01-50](https://user-images.githubusercontent.com/63245456/203402173-a5e13ee6-34af-4846-b1ec-a3e7b40762fa.png)

</details>
<details>
<summary>Account settings tab for demo organization owner without email set</summary>

![Screenshot from 2022-11-22 20-02-06](https://user-images.githubusercontent.com/63245456/203402335-199fb11f-3a35-428e-8c43-933855d44bea.png)

</details>

alya commented:
Nice!

For the button text, maybe something like this?

[text above the button]
Configure your email to invite other users or convert to a permanent Zulip organization.

Button: Configure email

alya commented:
What does "Deactivate account" do when the email hasn't been configured yet?

laurynmm commented:
> What does "Deactivate account" do when the email hasn't been configured yet? Maybe we should drop it?

As the owner is the only user in the Zulip organization, there will be an error if they click confirm on the deactivate account modal. Since this behavior doesn't change for a permanent account, I don't think we have to remove it. But we can if we think it would be confusing for the user. Here's a screenshot of the error in this context and normal context ...

<details>
<summary>Deactivate account demo organization - error for only one user</summary>

![Screenshot from 2022-11-30 19-36-08](https://user-images.githubusercontent.com/63245456/204880747-3897ef33-e098-4951-9ed4-868727873570.png)
</details>
<details>
<summary>Deactivate account permanent organization (current main) - error for only one user</summary>

![Screenshot from 2022-11-30 19-42-01](https://user-images.githubusercontent.com/63245456/204881821-a62c403c-98c0-43e3-8686-1d9bdd27ae64.png)
</details>

---

Here's the updated screenshot with the updated email configure text. I added links to the help center for inviting users and for making the demo organization permanent. 

<details>
<summary>Account settings tab - no email configured</summary>

![Screenshot from 2022-11-30 19-10-39](https://user-images.githubusercontent.com/63245456/204880386-7240480e-48b5-4d76-bc8e-991965550d16.png)
</details>

---

The modal for configuring the email doesn't 'work' yet, but I have the visual modeled on the change email form. Here's a current screenshot for feedback. I think there was conversation about making the confirm button disabled until the user makes a change to their email ... yeah, I opened #22683 for that. I don't know if we want to work on that in tandem to this.

<details>
<summary>Configure demo organization owner email modal</summary>

![Screenshot from 2022-11-30 19-21-19](https://user-images.githubusercontent.com/63245456/204881037-393eb0fc-d27d-4139-8148-e55b1edafcfb.png)
</details>

alya commented:
> Deactivate account demo organization - error for only one user

Ah, we should just disable the Deactivate account button in general when the action is not allowed. What is the rule? You can't deactivate your account if you're the only owner? Anyway, we should have a tooltip explaining why the button is disabled.

In general, we want to disable buttons for actions that are not permitted, rather than giving an error message after the button is pressed.

@laurynmm or @sahil839 , would one of you be up for taking care of this?

alya commented:
> yeah, I opened https://github.com/zulip/zulip/issues/22683 for that. I don't know if we want to work on that in tandem to this.

It would be nice, but not a blocker, I guess. Is there a reason that issue is not marked as "help wanted"?

alya commented:
Can we make sure the email modal does whatever email format validation we normally do at registration time?

laurynmm commented:
> Ah, we should just disable the Deactivate account button in general when the action is not allowed. What is the rule? You can't deactivate your account if you're the only owner? Anyway, we should have a tooltip explaining why the button is disabled.
> 
> In general, we want to disable buttons for actions that are not permitted, rather than giving an error message after the button is pressed.

In the majority of cases, the user will be able to deactivate their account. The two exceptions are if they are the only owner and if they are the only user/owner.

And I think this is a bit complicated to do on the frontend because the client (in this case the web-app) doesn't know that you are the **only** owner and/or the **only** user in the Zulip organization. I imagine that's why the button isn't deactivated.

I can start a #frontend discussion on CZO about this.

> > yeah, I opened #22683 for that. I don't know if we want to work on that in tandem to this.
> 
> It would be nice, but not a blocker, I guess. Is there a reason that issue is not marked as "help wanted"?

I'm not sure. Not to my knowledge.

laurynmm commented:
**Updates**:
- Added a first draft of the email configuration flow.
  - There's a separate email sent for confirming the email and when that confirmation happens, the user is redirected to setting a password for the account.
  - There is still fine tuning to do here, but this seemed like a good start.
- Disabled a number of settings before the email is configured.
  - Still to-do is the invite users modal.

@alya - If you get a chance before you're off, it'd be great to get your feedback on the updates. There's still some refactoring / clean-up I need to do on the code updates before it's ready for a technical review, but getting you're input on the text / flow of the email configuration and the disabled hints/tooltips would be great.

alya commented:
Thanks! A few thoughts on the UI (outside of emails):

1. "Configure email" sounds a bit confusing / overly technical to me. How about using "Add" instead of "Configure" everywhere (the line above the button, the button, the confirmation in the modal)?

2. This notice seems fine to me as-is, but just want to call it out explicitly in case someone has other thoughts:

>  Check your email (foo@bar.com) to confirm the new address.

3. On the password confirmation page, I wonder if we should make the wording work better for the case when there is no password yet? (I think this can also come up with SSO logins, not just for demo orgs.)

E.g.:
- We've reset your password! -> We've set a new password!
- Password successfully reset (page title) ->  New password successfully set

alya commented:
I think should also strongly encourage (or maybe even require) the user to configure their name before inviting others. Otherwise, the invitation email will be weird/confusing.

I'm open to different ways of doing this, but it would be nice to minimize disruptions in actually completing the invite flow. One UI that might be convenient is if, upon clicking the "Invite" button in the "Invite users" modal, we popped up a modal encouraging the user to change their name right there if they haven't done so yet.

alya commented on templates/zerver/emails/confirm_demo_organization_email.subject.txt:
```diff
@@ -0,0 +1 @@
+{{ _("Verify email address for demo organization owner") }}
```
I think we could just use `confirm_new_email_subject.txt` here without customizing.

alya commented:
For the email subject, maybe the following would work well?

1. For non-demo orgs, change it to: "Verify your new email address for `<organization name>`"
2. For demo orgs, change it to: "Verify your new email address for your demo Zulip organization`

alya commented:
For the email body, maybe something like this for the part above the button?

```
<p>{% trans realm_uri=macros.link_tag(realm_uri),
new_email=macros.email_tag(new_email) %}We received a request to add the email
address {{ new_email }} to your Zulip demo organization account on {{ realm_uri
}}. To confirm this update and set a password for this account, please click
below:{% endtrans %}
```

(and corresponding changes on the .txt version)

laurynmm commented:
**Updates**:
- Added a commit to disable clicking on the invite users links (until email is set).
  - Disables the links in the right sidebar and in the invitations tab of the organization settings, and adds a tooltip.
  - Removes the invite users link from the gear menu until the email is added.
  - This route seemed cleaner than allowing the user to open the modal with all the fields disabled, so I thought I would see what folks thoughts are.
- Updated the email flow for the review comments above. Specifically:
  - Changed all "Configure email" and "Configure" instances to "Add email" and "Add".
  - Updated email subjects (for both demo orgs and regular orgs), but I'll put up a separate PR for the regular org change if we decide we like it.
  - Updated email text for the above revisions.

**Still to-do**:
- Email validation as it appears on account creation vs on email change
- Having user update their name before sending links

---

**Screenshots**:

**Emails**:

<details>
<summary>Regular organization - user is changing their email - subject line is updated to include org name</summary>

![Screenshot from 2023-02-28 18-42-10](https://user-images.githubusercontent.com/63245456/221965985-af03c786-235b-4272-b7f3-0c52d9f4b3a0.png)
</details>
<details>
<summary>Demo organization - user is setting email - subject line and email text are both updated for revisions</summary>

![Screenshot from 2023-02-28 21-10-20](https://user-images.githubusercontent.com/63245456/221967355-7239100e-d5e7-4690-8e50-8674af24b886.png)
</details>

**Add email**:

<details>
<summary>Personal settings overlay - Account & privacy</summary>

![Screenshot from 2023-02-28 18-43-12](https://user-images.githubusercontent.com/63245456/221967824-23915189-5b73-49ca-99a8-cc627652bfb1.png)
</details>
<details>
<summary>Add email modal for demo organization</summary>

![Screenshot from 2023-02-28 18-43-28](https://user-images.githubusercontent.com/63245456/221967978-782ee2b1-8910-4115-937b-5e8e1cb5ca49.png)
</details>

**Invite users changes**:

<details>
<summary>Right sidebar</summary>

![Screenshot from 2023-02-28 20-07-00](https://user-images.githubusercontent.com/63245456/221968230-2521e652-96cd-4f57-ab34-24c48d625581.png)
</details>
<details>
<summary>Invitations tab in org settings overlay</summary>

![Screenshot from 2023-02-28 20-07-40](https://user-images.githubusercontent.com/63245456/221968381-67b55ac1-a0a9-4c01-9c58-1f93a249c6e5.png)
</details>
<details>
<summary>Gear menu - Invite users has been removed</summary>

![Screenshot from 2023-02-28 20-07-16](https://user-images.githubusercontent.com/63245456/221968509-e0ab7f79-8b6d-4703-a4ac-a3b9167c91f1.png)

</details>

alya commented:
Nice! A couple of comments:

1. I'm worried about the tooltip approach, because I really don't want folks to get stuck when trying to take the critical step of inviting users. We could try giving specific instructions in the tooltip (e.g., "To invite more users, go to Personal settings > Account & privacy, and add your email address."), but it would be much better if they could just click a link or two to get where they need to be. It's a bit of a tricky flow as it is. I dunno if we could make an interaction work where clicking on the "Invite more users" link would actually take you to Account & privacy, without it being too confusing. 

2. Small question: Does "Confirm and set password" always spill on two lines, or just in narrow screens?

I haven't done manual testing on this version.

laurynmm commented:
>   1. I'm worried about the tooltip approach, because I really don't want folks to get stuck when trying to take the critical step of inviting users. We could try giving specific instructions in the tooltip (e.g., "To invite more users, go to Personal settings > Account & privacy, and add your email address."), but it would be much better if they could just click a link or two to get where they need to be. It's a bit of a tricky flow as it is. I dunno if we could make an interaction work where clicking on the "Invite more users" link would actually take you to Account & privacy, without it being too confusing.

I've started a conversation about this in [CZO in the demo organization topic in frontend](https://chat.zulip.org/#narrow/stream/6-frontend/topic/demo.20organizations/near/1520902).

>   2. Small question: Does "Confirm and set password" always spill on two lines, or just in narrow screens?

Yes, I assume this is for the button in the email. It's currently set to a 200px max-width and that text just exceeds that value. So it will be on two lines.

I just pushed an update to clear the "has conflicts" label. There are no other updates/changes at this point.

I also started a conversation in frontend on CZO [here](https://chat.zulip.org/#narrow/stream/6-frontend/topic/email.20vaildation/near/1520842) about the email validation question.

laurynmm commented:
**Updates**:
- The invite user modal is disabled when the demo organization owner hasn't set an email for their account.
- Added `full_name` field to modal for adding the demo organization owner's email address.
  - This is to address encouraging the owner to update their name before inviting users.
  - The modal checks for an empty string for the email address in the case that the use tries to only update their name in the form.

@sahil839 - I was wondering if you'd be up for doing a review of the code changes so far since you've recently worked on the new organization creation forms/flow?

sahil839 commented on zerver/views/user_settings.py:
```diff
@@ -88,25 +90,40 @@ def confirm_email_change(request: HttpRequest, confirmation_key: str) -> HttpRes
+        # an email for the owner's account.
+        reset_password_url = generate_password_reset_url(user_profile, default_token_generator)
+        return HttpResponseRedirect(reset_password_url)
+    else:
```
We can remove this `else` since we already return above in the if-block.

sahil839 reviewed (commented):
Posted one minor comment about code changes. A couple of things that I noticed while manually testing the PR -
- I cannot see tooltip sometimes when hovering over email-address-visibility dropdown and API key button in demo org. Also, I guess we can remove the fading behavior on hovering over API key button when it is disabled.

![Peek 2023-07-24 12-36](https://github.com/zulip/zulip/assets/35494118/d04696e8-45ef-45fb-a32e-bbbd5a5ac5b7)

    
- We do not allow user keeping the name field empty while setting email in a demo org, but the button is enabled when I make the name field empty. Also the button is enabled, when I just remove a character from name field but email is empty. I don't think it is super important since we atleast show errors for these cases, but we can also keep the button disabled atleast when either one of two inputs is empty.

sahil839 reviewed (commented):
The submit button can still be enabled if one of the input is empty. I posted a comment with a proposed way to fix it. Looks good otherwise.

I also tested the hover behavior and tooltips, and it works fine.

laurynmm commented on web/src/settings_account.js:
```diff
@@ -743,6 +744,98 @@ export function set_up() {
+                $demo_organization_submit_button.prop(
+                    "disabled",
+                    $add_name_element.val().trim() === "",
+                );
```
Yeah, that makes much more sense. I've updated for the suggested selector / validation above.

laurynmm commented:
**Updates**:
- Updated for review feedback above.

**Screencast**:

<details>
<summary>Demo organization add email form</summary>

[Screencast from 2023-07-31 11-55-35.webm](https://github.com/zulip/zulip/assets/63245456/4ba57260-14e5-434d-8cd3-8875679128a4)
</details>

sahil839 commented:
Looks good to me.

timabbott commented on zerver/views/development/registration.py:
```diff
@@ -94,6 +94,7 @@ def register_demo_development_realm(request: HttpRequest) -> HttpResponse:
     realm_name = generate_demo_realm_name()
     realm_type = Realm.ORG_TYPES["business"]["id"]
     realm_subdomain = realm_name
+    email_address_visibility = UserProfile.EMAIL_ADDRESS_VISIBILITY_NOBODY
```
We probably will ultimately want to do this in `zerver/actions` somewhere, since I don't think the intent is to use this function for demo org creation in production.

laurynmm replied:
This is a field that is configured by users (or the org default) when a person registers for a Zulip organization now.

So, I was thinking that, when we update the landing page (`/new`) to be able to create a demo organization, we would have this set in the form/button that the user clicks.

timabbott replied:
OK, we can probably do it that way.

timabbott commented on web/src/tippyjs.js:
```diff
@@ -304,6 +304,18 @@ export function initialize() {
+        target: ["#api_key_button_container.disabled_setting_tooltip"],
+        content: $t({
+            defaultMessage:
+                "Demo organization owners must configure their email to access this feature.",
```
This message seems not directly correctly; the person seeing this will be doing a demo, and probably something like "The Zulip API is only available to accounts with an email address configured."

timabbott replied:
Or maybe something more like "You must configure your email address to access this feature" if we want it to be generic; my main concern here is about "Demo organization owners" being a bit too jargony.

timabbott commented on zerver/actions/user_settings.py:
```diff
@@ -152,15 +152,33 @@ def do_start_email_change_process(user_profile: UserProfile, new_email: str) ->
+            language=language,
+            context=context,
+            realm=user_profile.realm,
+        )
```
If the only thing different in this code path is the template name, we can probably make that a variable and be the only thing inside the `if/else` block?

timabbott commented:
@laurynmm thanks for all the work on this! I posted a handful of hopefully fairly surface level changes to consider; this branch otherwise LGTM.

laurynmm commented:
**Updates**:
- Updated disabled button tool tips and warning banners for a more generic message about adding/configuring email, removes "Demo organization owner" jargon. See screenshots below.
- Updated `zerver/actions/user_settings.py` to just update the email template as a variable for the blank email / demo organization owner case.

**Potential follow-up**:
- I was thinking we could update the demo organization help article to have a secttion with instructions for adding/configuring an email address, which we could then link to in the warning banners for the invite user modal and the authentication methods.

**Updated screenshots - all before user configures an email address**:

<details>
<summary>Account & privacy panel</summary>

**Note the tooltip is the same for the API key button and the email privacy setting dropdown.**
![Screenshot from 2023-08-31 12-30-05](https://github.com/zulip/zulip/assets/63245456/855af765-b174-4674-86a5-5cedc6ac7544)
</details>
<details>
<summary>Invite users modal</summary>

![Screenshot from 2023-08-31 12-30-34](https://github.com/zulip/zulip/assets/63245456/cea05bfc-a465-4d29-997c-8ee0af1f3ee4)
</details>
<details>
<summary>Authentication methods panel</summary>

![Screenshot from 2023-08-31 12-30-54](https://github.com/zulip/zulip/assets/63245456/0e0c38b5-8cd4-4b6e-888a-4bf1025c8d47)
</details>

timabbott commented:
> I was thinking we could update the demo organization help article to have a section with instructions for adding/configuring an email address, which we could then link to in the warning banners for the invite user modal and the authentication methods.

Yeah that sounds good.

timabbott commented on web/templates/invite_user_modal.hbs:
```diff
@@ -2,6 +2,9 @@
     <div class="alert" id="dev_env_msg"></div>
     {{/if}}
+    {{#unless user_has_email_set }}
+    <div class="tip">{{t "You must add an email address to your account to be able to invite users to Zulip." }}</div>
```
I wonder if this could just use the same "to access this feature" language we use in the settings pages? Mostly just thinking about extra work for translators.

timabbott commented:
Merged, with a few notes on follow-ups, thanks @laurynmm!
